kotlinx-coroutines-core/wasm/src/Promise.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import kotlin.coroutines.*
+import kotlin.js.*
+
+@Suppress("UNUSED_PARAMETER")
+internal fun promiseSetDeferred(promise: Promise<JsAny?>, deferred: JsAny): Unit =
+    js("promise.deferred = deferred")
+
+@Suppress("UNUSED_PARAMETER")
+internal fun promiseGetDeferred(promise: Promise<JsAny?>): JsAny? = js("""{
+    console.assert(promise instanceof Promise, "promiseGetDeferred must receive a promise, but got ", promise);
+    return promise.deferred == null ? null : promise.deferred; 
+}""")
+
+
+/**
+ * Starts new coroutine and returns its result as an implementation of [Promise].
+ *
+ * Coroutine context is inherited from a [CoroutineScope], additional context elements can be specified with [context] argument.
+ * If the context does not have any dispatcher nor any other [ContinuationInterceptor], then [Dispatchers.Default] is used.
+ * The parent job is inherited from a [CoroutineScope] as well, but it can also be overridden
+ * with corresponding [context] element.
+ *
+ * By default, the coroutine is immediately scheduled for execution.
+ * Other options can be specified via `start` parameter. See [CoroutineStart] for details.
+ *
+ * @param context additional to [CoroutineScope.coroutineContext] context of the coroutine.
+ * @param start coroutine start option. The default value is [CoroutineStart.DEFAULT].
+ * @param block the coroutine code.
+ */
+public fun <T> CoroutineScope.promise(
```

Comment:
This looks like a copy-paste of the implementation from the JS module with slight adjustments.

Is it possible to share the implementation in the intermediate source-set like it's done for JVM+K/N?

Reply:
Yeah. kotlin js implementations uses js arrays which is not a good idea for wasm. I kept the native implementation and removed atomic reference.

Reply:
I do not see a good solution for this, because every declaration or body are difference (at least here). If you think for some declarations/files there is a good (or have good ideas) how to combine two implementations, please say.